node/validator_store.go

```diff
@@ -0,0 +1,540 @@
+	duplicateRequestSalt uint32
+}
+
+var _ ValidatorStore = &validatorStore{}
+
+func NewValidatorStore(
+	ctx context.Context,
+	logger logging.Logger,
+	config *Config,
+	timeSource func() time.Time,
+	ttl time.Duration,
+	registry *prometheus.Registry) (ValidatorStore, error) {
+
+	if !config.LittDBEnabled {
+		logger.Warn("WARNING: This node is running with littDB disabled. " +
+			"This is a deprecated mode of operation, and will not be supported in future versions.")
+	}
+
+	littDBPath := path.Join(config.DbPath, LittDBPath)
+	levelDBPath := path.Join(config.DbPath, LevelDBPath)
+	levelDBDeletionPath := path.Join(config.DbPath, LevelDBDeletionPath)
+
+	// If we previously made an attempt at deleting the levelDB database but it was interrupted, delete it now.
+	_, err := os.Stat(levelDBDeletionPath)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("failed to stat path %s: %v", levelDBDeletionPath, err)
+	}
+	if err == nil {
+		// The previous attempt at deleting the levelDB database was interrupted.
+		logger.Warnf("partial deletion of levelDB database detected at %s. Deleting.", levelDBDeletionPath)
+		err = os.RemoveAll(levelDBDeletionPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to delete path %s: %v", levelDBDeletionPath, err)
+		}
+	}
+
+	// Check to see which DBs currently have data on disk.
+	_, err = os.Stat(levelDBPath)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("failed to stat path %s: %v", levelDBPath, err)
+	}
+	levelDBExists := err == nil
+
+	_, err = os.Stat(littDBPath)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("failed to stat path %s: %v", littDBPath, err)
+	}
+	littDBExists := err == nil
+	if littDBExists && !config.LittDBEnabled {
+		return nil, fmt.Errorf("Unable to do backwards migration. Once enabled, littDB cannot be disabled.")
+	}
+
+	var littDB litt.DB
+	var chunkTable litt.Table
+	var headerTable litt.Table
+	var levelDB kvstore.TableStore
+
+	// If we are still running with levelDB, start it up.
+	if !config.LittDBEnabled || levelDBExists {
+		logger.Infof("Using levelDB at %s", levelDBPath)
+
+		levelDB, err = tablestore.Start(logger, &tablestore.Config{
+			Type:                       tablestore.LevelDB,
+			Path:                       &levelDBPath,
+			GarbageCollectionEnabled:   true,
+			GarbageCollectionInterval:  time.Duration(config.ExpirationPollIntervalSec) * time.Second,
+			GarbageCollectionBatchSize: 1024,
+			Schema: []string{
+				BatchHeaderTableName,
+				BlobCertificateTableName,
+				BundleTableName,
+				MigrationTableName},
+			MetricsRegistry:               registry,
+			LevelDBDisableSeeksCompaction: config.LevelDBDisableSeeksCompactionV2,
+			LevelDBSyncWrites:             config.LevelDBSyncWritesV2,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to create new tablestore: %w", err)
+		}
+	}
+
+	// Set up migration if necessary.
+	var migrationComplete time.Time
```

Comment:
Can you explain the logic behind `migrationComplete`? Having not yet reviewed the migration code below, it's confusing to me why this is needed

Reply:
```
	// Set up migration if necessary. The migrationComplete time is the time when all data in the old levelDB instance
	// has exceeded its TTL. If the TTL is 2 weeks, then the migrationComplete time will be 2 weeks from now. At that
	// moment, it becomes safe to permanently stop and delete the levelDB database.
```